mapiproxy/servers/default/asyncemsmdb/dcesrv_asyncemsmdb.c

```diff
@@ -0,0 +1,1230 @@
+
+   \return MAPISTORE_SUCCESS on success, otherwise MAPISTORE error
+ */
+static enum mapistore_error get_properties_mapistore(TALLOC_CTX *mem_ctx, struct asyncemsmdb_private_data *p,
+						     uint64_t folderId, struct SPropTagArray *properties,
+						     void **data_pointers, enum MAPISTATUS *retvals)
+{
+	uint32_t			contextID;
+	enum mapistore_error		retval;
+	void				*context_object;
+	char				*uri = NULL;
+	bool				soft_deleted;
+	struct mapistore_property_data	*prop_data;
+	int				i;
+
+	/* Sanity checks */
+	MAPISTORE_RETVAL_IF(!p, MAPISTORE_ERR_NOT_INITIALIZED, NULL);
+	MAPISTORE_RETVAL_IF(!properties, MAPISTORE_ERR_INVALID_PARAMETER, NULL);
+	MAPISTORE_RETVAL_IF(!data_pointers, MAPISTORE_ERR_INVALID_PARAMETER, NULL);
+	MAPISTORE_RETVAL_IF(!retvals, MAPISTORE_ERR_INVALID_PARAMETER, NULL);
+
+	/* Retrieve the mapistore folder URI */
+	retval = mapistore_indexing_record_get_uri(p->mstore_ctx, p->username, mem_ctx, folderId, &uri, &soft_deleted);
+	MAPISTORE_RETVAL_IF(retval, retval, NULL);
+
+	/* Set or add context */
+	retval = mapistore_search_context_by_uri(p->mstore_ctx, uri, &contextID, &context_object);
+	if (retval == MAPISTORE_SUCCESS) {
+		retval = mapistore_add_context_ref_count(p->mstore_ctx, contextID);
+	} else {
+		retval = mapistore_add_context(p->mstore_ctx, p->username, uri, folderId, &contextID, &context_object);
+	}
+	MAPISTORE_RETVAL_IF(retval, retval, NULL);
+
+	prop_data = talloc_array(NULL, struct mapistore_property_data, properties->cValues);
+	MAPISTORE_RETVAL_IF(!prop_data, MAPISTORE_ERR_NO_MEMORY, NULL);
+	memset(prop_data, 0, sizeof(struct mapistore_property_data) * properties->cValues);
+
+	retval = mapistore_properties_get_properties(p->mstore_ctx, contextID, context_object, mem_ctx, properties->cValues,
+						     properties->aulPropTag, prop_data);
+	MAPISTORE_RETVAL_IF(retval, retval, prop_data);
+
+	for (i = 0; i < properties->cValues; i++) {
+		if (prop_data[i].error) {
+				if (properties->aulPropTag[i] == PidTagFolderFlags) {
+					prop_data[i].data = talloc_zero(prop_data, uint32_t);
+					*((uint32_t *)prop_data[i].data) = FolderFlags_IPM|FolderFlags_Normal;
+					data_pointers[i] = prop_data[i].data;
+					(void) talloc_reference(data_pointers, prop_data[i].data);
+					retvals[i] = MAPI_E_SUCCESS;
+				} else {
+					retvals[i] = mapistore_error_to_mapi(prop_data[i].error);
+				}
+		} else {
+			if (prop_data[i].data == NULL) {
+				if (properties->aulPropTag[i] == PidTagFolderFlags) {
+					prop_data[i].data = talloc_zero(prop_data, uint32_t);
+					*(uint32_t *)prop_data[i].data = FolderFlags_IPM|FolderFlags_Normal;
+					data_pointers[i] = prop_data[i].data;
+					(void) talloc_reference(data_pointers, prop_data[i].data);
+					retvals[i] = MAPI_E_SUCCESS;
+				} else {
+					retvals[i] = MAPI_E_NOT_FOUND;
+				}
+			} else {
+				if (properties->aulPropTag[i] == PidTagParentFolderId) {
+					if (*(uint64_t *)prop_data[i].data == 0x0) {
```

Comment:
I don't think it is worth it. As you already mentioned, there is a detailed explanation about the underlying reasoning. Therefore it is just about following the code logic to figure this out, which is what we expect from any developer.